File: .NET/VS2010TrainingKit/Labs/WhatsNewInWCF4/Source/Ex7-DiscoveryProxy/End/C#/DiscoveryChat/ChatWindow.cs
// ----------------------------------------------------------------------------------
// Microsoft Developer & Platform Evangelism
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
// 
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
// OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
// ----------------------------------------------------------------------------------
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
// ----------------------------------------------------------------------------------

namespace Microsoft.Samples.Discovery
{
    using System;
    using System.ComponentModel;
    using System.Globalization;
    using System.ServiceModel;
    using System.Windows.Forms;
    using Microsoft.Samples.Discovery.Contracts;
    using Microsoft.Samples.Discovery.Properties;

    public partial class ChatWindow : Form
    {
        private ISimpleChatService chatClient;
        private SimpleChat simpleChatOwner;
        private PeerUser peerUser;

        public ChatWindow(SimpleChat simpleChat, PeerUser peerUserObject)
        {
            if (simpleChat == null)
            {
                throw new ArgumentNullException("simpleChat");
            }

            if (peerUserObject == null)
            {
                throw new ArgumentNullException("peerUserObject");
            }

            InitializeComponent();

            this.chatClient =
                ChannelFactory<ISimpleChatService>.CreateChannel(
                new WSHttpBinding(), peerUserObject.Address);

            this.simpleChatOwner = simpleChat;

            this.peerUser = peerUserObject;

            string chatWith = null;

            if (!string.IsNullOrEmpty(peerUserObject.UserName))
            {
                chatWith = this.peerUser.UserName;
            }
            else
            {
                chatWith = this.peerUser.Address.Uri.Host;
            }

            Text = string.Format(
                CultureInfo.CurrentCulture,
                Resources.ChattingWithMessage,
                simpleChat.UserName,
                chatWith);
            Owner = simpleChat;

            StartPosition = FormStartPosition.CenterParent;
            Show();
            chatText.Focus();
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Demo code")]
        private void SendMessage_Click(object sender, EventArgs e)
        {
            try
            {
                ChatMessage sentMsg = new ChatMessage(this.simpleChatOwner.UserName, chatText.Text, this.simpleChatOwner.LocalAddress);
                this.chatClient.ProcessSimpleMessage(sentMsg);
                this.DisplayMessage(sentMsg);
                this.chatText.Clear();
            }
            catch (CommunicationException cex)
            {
                ((ICommunicationObject)chatClient).Abort();
                MessageBox.Show(
                    string.Format(
                        CultureInfo.CurrentCulture,
                        Resources.UserNotAvailableMessage,
                        this.peerUser.UserName,
                        cex.Message),
                        this.Text,
                        MessageBoxButtons.OK,
                        MessageBoxIcon.Information,
                        MessageBoxDefaultButton.Button1,
                        (MessageBoxOptions)0);
                this.simpleChatOwner.RemoveUser(this.peerUser.Address.Uri);
                this.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show(
                    string.Format(CultureInfo.CurrentCulture, Resources.UnknownErrorMessage, ex.Message),
                        this.Text,
                        MessageBoxButtons.OK,
                        MessageBoxIcon.Information,
                        MessageBoxDefaultButton.Button1,
                        (MessageBoxOptions)0);
            }
        }

        private void ChatText_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                e.SuppressKeyPress = true;
                this.SendMessage_Click(sender, e);
            }
        }

        protected override void OnClosing(CancelEventArgs e)
        {
            if (this.chatClient != null)
            {
                try
                {
                    ((ICommunicationObject)chatClient).Close();
                }
                catch (CommunicationException)
                {
                    // No worries, closing
                }
            }

            base.OnClosing(e);
        }

        public void DisplayMessage(ChatMessage message)
        {
            if (message == null)
            {
                throw new ArgumentNullException("message");
            }

            chatBox.Items.Add(message.UserName + " says: " + message.Message);
        }
    }
}

File: .NET/VS2010TrainingKit/Labs/WhatsNewInWCF4/Source/Ex7-DiscoveryProxy/End/VB/DiscoveryChat/ChatWindow.vb
' ----------------------------------------------------------------------------------
' Microsoft Developer & Platform Evangelism
' 
' Copyright (c) Microsoft Corporation. All rights reserved.
' 
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
' OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
' ----------------------------------------------------------------------------------
' The example companies, organizations, products, domain names,
' e-mail addresses, logos, people, places, and events depicted
' herein are fictitious.  No association with any real company,
' organization, product, domain name, email address, logo, person,
' places, or events is intended or should be inferred.
' ----------------------------------------------------------------------------------

Imports System.ComponentModel
Imports System.Globalization
Imports System.ServiceModel
Imports Microsoft.Samples.Discovery.Contracts

Partial Public Class ChatWindow
    Inherits Form
    Private chatClient As ISimpleChatService
    Private simpleChatOwner As SimpleChat
    Private peerUser As PeerUser

    Public Sub New(ByVal simpleChat As SimpleChat, ByVal peerUserObject As PeerUser)
        If simpleChat Is Nothing Then Throw New ArgumentNullException("simpleChat")

        If peerUserObject Is Nothing Then Throw New ArgumentNullException("peerUserObject")

        InitializeComponent()

        Me.chatClient = ChannelFactory(Of ISimpleChatService).CreateChannel(New WSHttpBinding(), peerUserObject.Address)

        Me.simpleChatOwner = simpleChat

        Me.peerUser = peerUserObject

        Dim chatWith As String = Nothing

        If (Not String.IsNullOrEmpty(peerUserObject.UserName)) Then
            chatWith = Me.peerUser.UserName
        Else
            chatWith = Me.peerUser.Address.Uri.Host
        End If

        Text = String.Format(CultureInfo.CurrentCulture, My.Resources.ChattingWithMessage, simpleChat.UserName, chatWith)
        Owner = simpleChat

        StartPosition = FormStartPosition.CenterParent
        Show()
        chatText.Focus()
    End Sub

    <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification:="Demo code")>
    Private Sub SendMessage_Click() Handles sendMessage.Click
        Try
            Dim sentMsg As New ChatMessage(Me.simpleChatOwner.UserName, chatText.Text, Me.simpleChatOwner.LocalAddress)
            Me.chatClient.ProcessSimpleMessage(sentMsg)
            Me.DisplayMessage(sentMsg)
            Me.chatText.Clear()
        Catch cex As CommunicationException
            CType(chatClient, ICommunicationObject).Abort()
            MsgBox(String.Format(CultureInfo.CurrentCulture, My.Resources.UserNotAvailableMessage, Me.peerUser.UserName, cex.Message), MsgBoxStyle.Information, Me.Text)
            Me.simpleChatOwner.RemoveUser(Me.peerUser.Address.Uri)
            Me.Close()
        Catch ex As Exception
            MsgBox(String.Format(CultureInfo.CurrentCulture, My.Resources.UnknownErrorMessage, ex.Message), MsgBoxStyle.Information, Me.Text)
        End Try
    End Sub

    Private Sub ChatText_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles chatText.KeyDown
        If e.KeyCode = Keys.Enter Then
            e.SuppressKeyPress = True
            Me.SendMessage_Click()
        End If
    End Sub

    Protected Overrides Sub OnClosing(ByVal e As CancelEventArgs)
        If Me.chatClient IsNot Nothing Then
            Try
                CType(chatClient, ICommunicationObject).Close()
            Catch e1 As CommunicationException
                ' No worries, closing
            End Try
        End If

        MyBase.OnClosing(e)
    End Sub

    Public Sub DisplayMessage(ByVal message As ChatMessage)
        If message Is Nothing Then Throw New ArgumentNullException("message")

        chatBox.Items.Add(message.UserName & " says: " & message.Message)
    End Sub
End Class